gpflow/likelihoods.py

```diff
@@ -583,3 +583,132 @@ def conditional_variance(self, F):
+
+        else:
+            def eval_func(func):
+                return tf.reshape(func(mc_Xr),
+                                  (self.num_monte_carlo_points, N, D))
+
+        if isinstance(funcs, Iterable):
+            return [eval_func(f) for f in funcs]
+        else:
+            return eval_func(funcs)
+
+    def predict_mean_and_var(self, Fmu, Fvar, epsilon=None):
+        r"""
+        Given a Normal distribution for the latent function,
+        return the mean of Y
+
+        if
+            q(f) = N(Fmu, Fvar)
+
+        and this object represents
+
+            p(y|f)
+
+        then this method computes the predictive mean
+
+           \int\int y p(y|f)q(f) df dy
+
+        and the predictive variance
+
+           \int\int y^2 p(y|f)q(f) df dy  - [ \int\int y^2 p(y|f)q(f) df dy ]^2
+
+        Here, we implement a default Monte Carlo routine.
+        """
+        integrand2 = lambda *X: self.conditional_variance(*X) + tf.square(self.conditional_mean(*X))
+        mc_y, mc_y2 = self._mc_evals([self.conditional_mean, integrand2],
+                                     Fmu, Fvar, epsilon=epsilon)
+        E_y = tf.reduce_mean(mc_y, axis=0)
+        E_y2 = tf.reduce_mean(mc_y2, axis=0)
+        V_y = E_y2 - tf.square(E_y)
+        return E_y, V_y  # N x D
+
+    def predict_density(self, Fmu, Fvar, Y, epsilon=None):
+        r"""
+        Given a Normal distribution for the latent function, and a datum Y,
+        compute the (log) predictive density of Y.
+
+        i.e. if
+            q(f) = N(Fmu, Fvar)
+
+        and this object represents
+
+            p(y|f)
+
+        then this method computes the predictive density
+
+           \int p(y=Y|f)q(f) df
+
+        Here, we implement a default Monte Carlo routine.
+        """
+        mc_logp = self._mc_evals(self.logp, Fmu, Fvar, Y=Y, epsilon=epsilon)
+        log_N = tf.log(tf.cast(self.num_monte_carlo_points, settings.float_type))
+        return tf.reduce_logsumexp(mc_logp, axis=0) - log_N  # N x 1
+
+    def variational_expectations(self, Fmu, Fvar, Y, epsilon=None):
+        r"""
+        Compute the expected log density of the data, given a Gaussian
+        distribution for the function values.
+
+        if
+            q(f) = N(Fmu, Fvar)  - Fmu: N x D  Fvar: N x D
+
+        and this object represents
+
+            p(y|f)  - Y: N x 1
+
+        then this method computes
+
+           \int (\log p(y|f)) q(f) df.
+
+
+        Here, we implement a default Monte Carlo quadrature routine.
+        """
+        mc_logp = self._mc_evals(self.logp, Fmu, Fvar, Y=Y, epsilon=epsilon)
+        return tf.reduce_mean(mc_logp, axis=0)  # N x 1
+
+
+class GaussianMC(MonteCarloLikelihood, Gaussian):
+    """
+    Stochastic version of Gaussian likelihood for comparison.
+    """
+    def __init__(self, variance=1.0, name=None):
+        super().__init__(name=name)
+        self.variance = variance
```

Comment:
Not that I really care about this, but this looks like a perfect example of thinking about cooperative multiple inheritance. Should the constructor of `MonteCarloLikelihood` not also call `super()` with kwargs?

If you're interested, check out https://www.youtube.com/watch?v=EiOglTERPEo. It's a good watch anyway.

Reply:
Like in 69f4c59 ? :)